### PR TITLE
Set output batch rows adaptively for operators

### PR DESCRIPTION
**Pull request by gggrace14**
Add the outputBatchRows() to Operator interface to determine the number
of output batch rows based on the total batch size config
preferred_output_batch_bytes and the given average row size.
If the average row size is not given, return the config
preferred_output_batch_rows.

Make the HashAggregation, OrderBy and Windown operator use the above logic
to adaptively set the number of output batch rows based on the estimated
output row size.

### CODE REVIEW

**Comment by gggrace14** [comment]
Hi @xiaoxmeng , this change breaks two more unit tests in AggregationTest.cpp, spillWithEmptyPartition, spillWithNonSpillingPartition, in addition to the one we discussed. The tests end with exceptions, like P672036745, P672037418. It seems to come from MemoryUsageTracker::incrementReservation(). Would you be able to help fix?

(I spent some time to look into, but feel really difficult for me to understand it. It's not impossible that my code returns an invalid value that breaks the MemoryUsageTracker)

File: velox/core/QueryConfig.h [hunk 1]
```diff
@@ -97,10 +97,15 @@ class QueryConfig {
   static constexpr const char* kMaxPartitionedOutputBufferSize =
       "driver.max-page-partitioning-buffer-size";
 
-  /// Preffered number of rows to be returned by operators from
+  /// Preferred size of batches in bytes to be returned by operators from
```
Comment (mbasmanova) [inline velox/core/QueryConfig.h:100]: Let's update the docs: https://facebookincubator.github.io/velox/configs.html

File: velox/core/QueryConfig.h [hunk 2]
```diff
@@ -97,10 +97,15 @@ class QueryConfig {
...
   /// Operator::getOutput.
-  static constexpr const char* kPreferredOutputBatchSize =
-      "preferred_output_batch_size";
+  static constexpr const char* kPreferredOutputBatchBytes =
+      "preferred_output_batch_bytes";
+
+  /// Preferred number of rows to be returned by operators from
```
Comment (mbasmanova) [inline velox/core/QueryConfig.h:105]: The comment says "preferred", but the property name was changed to "default". It is actually confusing to use the name "default" because that property itself has a default value of 1024. Perhaps, keep the original name, but explain how the two properties are being used.

preferred_output_batch_bytes is used when operator can estimate row sizes.
preferred_output_batch_rows is used when operator cannot estimate row sizes.

File: velox/exec/GroupingSet.cpp [hunk 3]
```diff
@@ -694,4 +694,10 @@ void GroupingSet::updateRow(SpillMergeStream& input, char* FOLLY_NONNULL row) {
   mergeSelection_.setValid(input.currentIndex(), false);
 }
 
+int64_t GroupingSet::estimatedRowSize() const {
```
Comment (mbasmanova) [inline velox/exec/GroupingSet.cpp:697]: nit: perhaps, estimateRowSize (verb)

File: velox/exec/GroupingSet.h [hunk 4]
```diff
@@ -92,6 +92,9 @@ class GroupingSet {
     return table_ ? table_->rows()->numRows() : 0;
   }
 
+  /// Returns the estimated size of the row in memory.
```
Comment (mbasmanova) [inline velox/exec/GroupingSet.h:95]: perhaps, "Returns an estimate of the average row size."

File: velox/exec/Operator.h [hunk 5]
```diff
@@ -476,6 +478,13 @@ class Operator : public BaseRuntimeStatWriter {
   // 'identityProjections_' and 'resultProjections_'.
   RowVectorPtr fillOutput(vector_size_t size, BufferPtr mapping);
 
+  // Returns the max number of rows in the output batch.
+  virtual uint32_t outputBatchRows() const;
```
Comment (mbasmanova) [inline velox/exec/Operator.h:482]: Does this method need to be virtual?

Would you add some more details about how this method computes the result? I think it uses estimatedOutputRowSize to calculate how many rows fit in preferredOutputBatchBytes. It caps the number of rows at 10K and returns at least 1. If average row size is not know, it returns preferredOutputBatchRows.

File: velox/exec/Operator.h [hunk 6]
```diff
@@ -476,6 +478,13 @@ class Operator : public BaseRuntimeStatWriter {
...
+
+  // Estimated size of an output row. Returns the default value -1 if this
```
Comment (mbasmanova) [inline velox/exec/Operator.h:484]: Would it make sense to return an std::optional<int64_t> instead of using a special value for unknown?

File: velox/exec/Operator.h [hunk 7]
```diff
@@ -476,6 +478,13 @@ class Operator : public BaseRuntimeStatWriter {
...
+  // operator has no estimate.
+  virtual int64_t estimatedOutputRowSize() const;
```
Comment (mbasmanova) [inline velox/exec/Operator.h:486]: nit: estimateOutputRowSize (verb)

File: velox/exec/Operator.cpp [hunk 8]
```diff
@@ -294,6 +294,25 @@ OperatorStats Operator::stats(bool clear) {
...
+uint32_t Operator::outputBatchRows() const {
+  const auto& queryConfig = operatorCtx_->task()->queryCtx()->queryConfig();
+
+  auto rowSize = estimatedOutputRowSize();
+  if (rowSize == kUnknownRowSize) {
+    return queryConfig.defaultOutputBatchRows();
+  }
+
+  if (rowSize * 10'000 < queryConfig.preferredOutputBatchBytes()) {
+    return 10'000; // No more than 10'000 rows
```
Comment (mbasmanova) [inline velox/exec/Operator.cpp:306]: Should we make this a config as well?

File: velox/exec/Operator.cpp [hunk 9]
```diff
@@ -294,6 +294,25 @@ OperatorStats Operator::stats(bool clear) {
...
+  }
+  return std::max<uint32_t>(
+      queryConfig.preferredOutputBatchBytes() / rowSize, 1);
```
Comment (mbasmanova) [inline velox/exec/Operator.cpp:309]: Should we check if rowSize is zero or negative?

File: velox/exec/OperatorUtils.h [hunk 10]
```diff
@@ -19,7 +19,6 @@
 #include "velox/exec/Spiller.h"
 
 namespace facebook::velox::exec {
-
```
Comment (mbasmanova) [inline velox/exec/OperatorUtils.h:22]: Unrelated change?

File: velox/exec/OrderBy.cpp [hunk 11]
```diff
@@ -92,11 +92,7 @@ OrderBy::OrderBy(
 #endif
 
   outputBatchSize_ = std::max<uint32_t>(
```
Comment (mbasmanova) [inline velox/exec/OrderBy.cpp:94]: Why can't we just use outputBatchRows()?

File: velox/exec/RowContainer.h [hunk 12]
```diff
@@ -525,6 +525,10 @@ class RowContainer {
         stringAllocator_.freeSpace());
   }
 
+  // Returns the average size of rows in bytes stored in this container.
+  // Returns -1 if the container is empty.
+  int64_t estimatedRowSize() const;
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:530]: It would be nice to add a test for this logic to velox/exec/tests/RowContainerTest.cpp

File: velox/exec/tests/AggregationTest.cpp [hunk 13]
```diff
@@ -1417,5 +1433,59 @@ TEST_F(AggregationTest, preGroupedAggregationWithSpilling) {
...
+
+  // Test setting larger output batch bytes will create batches of greater
+  // number of rows.
+  {
+    auto outputBatchBytes = "1000";
+    auto task =
+        AssertQueryBuilder(plan, duckDbQueryRunner_)
+            .config(QueryConfig::kPreferredOutputBatchBytes, outputBatchBytes)
+            .config(
+                QueryConfig::kDefaultOutputBatchRows,
```
Comment (mbasmanova) [inline velox/exec/tests/AggregationTest.cpp:1461]: Why do we need to set this config? I assume it is not used in the aggregation operator. Is it?

**Review by mbasmanova — COMMENTED** [review]
@gggrace14 Overall looks good. Some comments below.

File: velox/exec/tests/AggregationTest.cpp [hunk 14]
```diff
@@ -1321,39 +1321,55 @@ TEST_F(AggregationTest, outputBatchSizeCheckWithSpill) {
...
+
+  auto plan = PlanBuilder()
+                  .values(batches)
+                  .singleAggregation({"c0", "c1"}, {"sum(c2)"})
+                  .planNode();
+  auto results = AssertQueryBuilder(plan).copyResults(pool_.get());
+
+  {
+    auto tempDirectory = exec::test::TempDirectoryPath::create();
+    uint64_t outputBufferSize = 10UL << 20;
```
Comment (xiaoxmeng) [inline velox/exec/tests/AggregationTest.cpp:1333]: Set outputBufferSize to a small value in the test failed test cases? thanks!

File: velox/exec/Operator.cpp (hunk 9 above)
Comment (gggrace14) [reply velox/exec/Operator.cpp:309]: You're right. Missed it.

File: velox/exec/Operator.h (hunk 5 above)
Comment (gggrace14) [reply velox/exec/Operator.h:482]: Thanks for the text!

File: velox/exec/tests/AggregationTest.cpp (hunk 13 above)
Comment (gggrace14) [reply velox/exec/tests/AggregationTest.cpp:1461]: Removed. You're right

File: velox/exec/RowContainer.h (hunk 12 above)
Comment (gggrace14) [reply velox/exec/RowContainer.h:530]: Added one unit test for it.

File: velox/exec/Operator.cpp (hunk 8 above)
Comment (gggrace14) [reply velox/exec/Operator.cpp:306]: @mbasmanova , I got this logic of capping # of rows by 10'000 from the TableScan::setBatchSize(). Orri once mentioned that this logic is in TableScan because # of row that TableScan can read once is no more than # of rows of a row group, 10'000 rows. Then in general case here shall we remove the cap?

File: velox/exec/OrderBy.cpp (hunk 11 above)
Comment (gggrace14) [reply velox/exec/OrderBy.cpp:94]: I'm changing to only use outputBatchRows().

File: velox/core/QueryConfig.h (hunk 2 above)
Comment (gggrace14) [reply velox/core/QueryConfig.h:105]: Revised

File: velox/core/QueryConfig.h (hunk 1 above)
Comment (gggrace14) [reply velox/core/QueryConfig.h:100]: Done

File: velox/exec/Operator.h [hunk 15]
```diff
@@ -476,6 +476,16 @@ class Operator : public BaseRuntimeStatWriter {
...
 
+  // Returns the max number of rows in the output batch. This uses
+  // estimateOutputRowSize to calculate how many rows fit in
+  // preferredOutputBatchBytes. It caps the number of rows at 10K and returns at
+  // least one row. If the operator does not have estimateOutputRowSize, returns
+  // preferredOutputBatchRows.
+  uint32_t outputBatchRows() const;
+
+  // Returns an estimate of the output row size if an operator has an estimate.
+  virtual std::optional<uint64_t> estimateOutputRowSize() const;
```
Comment (mbasmanova) [inline velox/exec/Operator.h:487]: I was thinking about this a bit more and wondering if we can avoid introducing a new virtual method, but rather pass the estimate to outputBatchRows():

```
uint32_t outputBatchRows(std::optional<uint64_t> estimatedRowSize = std::nullopt) const;
```

File: velox/exec/Operator.cpp [hunk 16]
```diff
@@ -294,6 +294,29 @@ OperatorStats Operator::stats(bool clear) {
...
+uint32_t Operator::outputBatchRows() const {
+  const auto& queryConfig = operatorCtx_->task()->queryCtx()->queryConfig();
+
+  if (!estimateOutputRowSize().has_value()) {
+    return queryConfig.defaultOutputBatchRows();
+  }
+
+  uint64_t rowSize = estimateOutputRowSize().value();
+
+  if (FOLLY_UNLIKELY(rowSize <= 0)) {
```
Comment (mbasmanova) [inline velox/exec/Operator.cpp:306]: Should we throw instead? VELOX_CHECK_GT(rowSize, 0)

File: velox/exec/RowContainer.h [hunk 17]
```diff
@@ -525,6 +525,10 @@ class RowContainer {
         stringAllocator_.freeSpace());
   }
 
+  // Returns the average size of rows in bytes stored in this container.
+  // Returns -1 if the container is empty.
+  int64_t estimateRowSize() const;
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:530]: Should we return std::optional instead of a special value -1?

File: velox/exec/RowContainer.cpp [hunk 18]
```diff
@@ -560,6 +561,16 @@ void RowContainer::extractProbedFlags(
...
 }
 
+int64_t RowContainer::estimateRowSize() const {
+  if (numRows_ == 0) {
+    return -1;
+  }
+  int64_t freeBytes = rows_.availableInRun() + fixedRowSize_ * numFreeRows_;
+  int64_t usedSize = rows_.allocatedBytes() - freeBytes +
+      stringAllocator_.retainedSize() - stringAllocator_.freeSpace();
+  return usedSize / numRows_;
```
Comment (mbasmanova) [inline velox/exec/RowContainer.cpp:571]: Can this produce a zero if usedSize < numRows_?

Is it possible for usedSize to be negative? Should we add VELOX_CHECK just in case?

File: velox/exec/tests/AggregationTest.cpp [hunk 19]
```diff
@@ -988,6 +988,7 @@ DEBUG_ONLY_TEST_F(AggregationTest, spillWithEmptyPartition) {
             .config(
                 QueryConfig::kSpillStartPartitionBit,
                 std::to_string(kPartitionStartBit))
+            .config(QueryConfig::kPreferredOutputBatchBytes, "1024")
```
Comment (mbasmanova) [inline velox/exec/tests/AggregationTest.cpp:991]: Do we need to specify this since it is the default?

File: velox/exec/tests/RowContainerTest.cpp [hunk 20]
```diff
@@ -729,6 +729,23 @@ TEST_F(RowContainerTest, rowSizeWithNormalizedKey) {
...
+  auto allRows = 100'000;
+  for (auto i = 0; i < allRows; i++) {
+    data->newRow();
+  }
+
+  RowContainerIterator iter;
+  char* rows[allRows];
+  // Test sizes smaller than the total row container size.
+  std::vector<int64_t> sizes = {1, 1 << 10, 1 << 20};
+  for (auto size : sizes) {
```
Comment (mbasmanova) [inline velox/exec/tests/RowContainerTest.cpp:743]: Would you explain how this test works? I'm surprised to see a call to listRows as I expect estimateRowSize to not be affected by a call to listRows. However, I'd expect estimateRowSize to depend on how much variable-size data has been written. It would be nice to write not only fixed-width rows, but also variable-width rows.

**Review by mbasmanova — COMMENTED** [review]
@gggrace14 Ge, thank you for iterating on this PR. Some follow-up comments.

File: velox/exec/Operator.cpp (hunk 16 above)
Comment (gggrace14) [reply velox/exec/Operator.cpp:306]: Okay with it. Revised.

File: velox/exec/RowContainer.h (hunk 17 above)
Comment (gggrace14) [reply velox/exec/RowContainer.h:530]: Revised.

File: velox/exec/RowContainer.cpp (hunk 18 above)
Comment (gggrace14) [reply velox/exec/RowContainer.cpp:571]: It makes sense. VELOX_CHECK is added. I originally thought leaving the caller to handle the edge cases.

File: velox/exec/tests/AggregationTest.cpp (hunk 19 above)
Comment (gggrace14) [reply velox/exec/tests/AggregationTest.cpp:991]: The default value of kPreferredOutputBatchBytes is 10MBs. Here is to set a small output batch size to remove a memory limit error.

File: velox/exec/Operator.h (hunk 15 above)
Comment (gggrace14) [reply velox/exec/Operator.h:487]: I got your idea. Revised.

File: velox/exec/tests/RowContainerTest.cpp (hunk 20 above)
Comment (gggrace14) [reply velox/exec/tests/RowContainerTest.cpp:743]: Sorry, it's a bit confusing. I renamed the variables, hopefully, to make it more readable.

listRows() doesn't have to do with estimateRowSize(). listRows() is used to get the fact or the expected numRows given the different batchSizes. Then EXPECT_EQ says the data->estimateRowSize() is accurate to derive the actual num of rows that matches the expected numRows.

File: velox/exec/Operator.h [hunk 21]
```diff
@@ -476,6 +476,13 @@ class Operator : public BaseRuntimeStatWriter {
   // 'identityProjections_' and 'resultProjections_'.
   RowVectorPtr fillOutput(vector_size_t size, BufferPtr mapping);
 
+  // Returns the number of rows for the output batch. This uses averageRowSize
+  // to calculate how many rows fit in preferredOutputBatchBytes. It caps the
+  // number of rows at 10K and returns at least one row. If the averageRowSize
+  // is not given, returns preferredOutputBatchRows.
+  uint32_t outputBatchRows(
+      const std::optional<uint64_t>& averageRowSize = std::nullopt) const;
```
Comment (mbasmanova) [inline velox/exec/Operator.h:484]: nit: pass by value

File: velox/exec/tests/RowContainerTest.cpp [hunk 22]
```diff
@@ -729,6 +729,26 @@ TEST_F(RowContainerTest, rowSizeWithNormalizedKey) {
...
+  RowContainerIterator iter;
+  char* result[unlimited];
+  // Batch sizes smaller than the total row container size.
+  std::vector<int64_t> batchSizes = {1, 1 << 10, 1 << 20};
+  for (auto batchSize : batchSizes) {
+    // Get the expected numRows based on the given batchSize.
+    auto numRows = data->listRows(&iter, unlimited, batchSize, result);
+    // Test estimateRowSize is accurate to get the expected numRows.
+    EXPECT_EQ(
+        std::ceil(batchSize * 1.0 / data->estimateRowSize().value()), numRows);
```
Comment (mbasmanova) [inline velox/exec/tests/RowContainerTest.cpp:748]: I still don't understand how this test verifies the behavior of data->estimateRowSize(). This behavior should not depend on the 'batchSize', hence, I don't see why logic around 'batchSizes' is relevant. At the same time, I'd expect this test to check scenarios where (1) data->estimateRowSize() returns std::nullopt; (2) data is variable-width and data->estimateRowSize() is taking that part of the data into account.

**Review by mbasmanova — COMMENTED** [review]
@gggrace14 Looks good to me % comment about RowContainerTest. Also, I wonder if we should use the new Operator::outputBatchRows in TableScan operator as well.

**Comment by gggrace14** [comment]
> @gggrace14 Looks good to me % comment about RowContainerTest. Also, I wonder if we should use the new Operator::outputBatchRows in TableScan operator as well.

Hi @mbasmanova , I have the comments addressed and revised the unit test for RowContainer. Would you be able to help take another look? I made TableScan use the new Operator::outputBatchRows, and still made the change in existing where your comments are kept but rebased it.

File: velox/docs/configs.rst [hunk 23]
```diff
@@ -2,6 +2,38 @@
...
+
+``preferred_output_batch_rows``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``integer``
+    * **Default value:** ``1024``
+
+Preferred number of rows to be returned by operators from Operator::getOutput. It is used
+when an estimate of average row size is not known. When the estimate of average row size is known,
+kPreferredOutputBatchBytes is used.
```
Comment (mbasmanova) [inline velox/docs/configs.rst:25]: same

File: velox/docs/configs.rst [hunk 24]
```diff
@@ -2,6 +2,38 @@
...
+
+``max_output_batch_rows``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``integer``
+    * **Default value:** ``10000``
+
+Max number of rows that could be return by operators from Operator::getOutput. It is used
+when an estimate of average row size is known and kPreferredOutputBatchBytes is used to compute
```
Comment (mbasmanova) [inline velox/docs/configs.rst:34]: same

File: velox/exec/OrderBy.cpp [hunk 25]
```diff
@@ -91,12 +91,7 @@ OrderBy::OrderBy(
   }
 #endif
 
-  outputBatchSize_ = std::max<uint32_t>(
-      operatorCtx_->execCtx()
-          ->queryCtx()
-          ->queryConfig()
-          .preferredOutputBatchSize(),
-      data_->estimatedNumRowsPerBatch(kBatchSizeInBytes));
+  outputBatchSize_ = outputBatchRows(data_->fixedRowSize());
```
Comment (mbasmanova) [inline velox/exec/OrderBy.cpp:94]: Why data_->fixedRowSize()?

File: velox/exec/RowContainer.cpp [hunk 26]
```diff
@@ -560,6 +561,19 @@ void RowContainer::extractProbedFlags(
...
 
+std::optional<int64_t> RowContainer::estimateRowSize() const {
+  if (numRows_ == 0) {
+    return std::nullopt;
+  }
+  int64_t freeBytes = rows_.availableInRun() + fixedRowSize_ * numFreeRows_;
+  int64_t usedSize = rows_.allocatedBytes() - freeBytes +
+      stringAllocator_.retainedSize() - stringAllocator_.freeSpace();
+  int64_t rowSize = usedSize / numRows_;
+  VELOX_CHECK_GE(
```
Comment (mbasmanova) [inline velox/exec/RowContainer.cpp:572]: Should this be GT ?

File: velox/exec/tests/RowContainerTest.cpp [hunk 27]
```diff
@@ -729,6 +729,37 @@ TEST_F(RowContainerTest, rowSizeWithNormalizedKey) {
...
+  auto numRows = 1000;
+
+  // Make a RowContainer with a fixed-length key column and a variable-length
+  // dependent column.
+  auto rowContainer = makeRowContainer({BIGINT()}, {VARCHAR()});
+  EXPECT_FALSE(rowContainer->estimateRowSize().has_value());
+
+  // Store rows to the container.
+  auto input = std::dynamic_pointer_cast<RowVector>(
+      BatchMaker::createBatch(ROW({BIGINT(), VARCHAR()}), numRows, *pool_));
```
Comment (mbasmanova) [inline velox/exec/tests/RowContainerTest.cpp:742]: BatchMaker is deprecated. Use VectorMaker instead.

File: velox/exec/tests/RowContainerTest.cpp [hunk 28]
```diff
@@ -729,6 +729,37 @@ TEST_F(RowContainerTest, rowSizeWithNormalizedKey) {
...
+  DecodedVector decodedDependent(*input->childAt(1), allRows);
+  for (size_t i = 0; i < numRows; i++) {
+    auto row = rowContainer->newRow();
+    rowContainer->store(decodedKey, i, row, 0);
+    rowContainer->store(decodedDependent, i, row, 1);
+  }
+
+  RowContainerIterator iter;
+  char* result[numRows];
+  auto resultRows = rowContainer->listRows(
```
Comment (mbasmanova) [inline velox/exec/tests/RowContainerTest.cpp:754]: What's the point of listing the rows? Shouldn't we call estimateRowSize instead?

**Review by mbasmanova — COMMENTED** [review]
@gggrace14 Ge, thank you for iterating on this PR. Some further comments.

**Review by mbasmanova — COMMENTED** [review]
@gggrace14 Ge, thank you for iterating on this PR. Some further comments.

File: velox/exec/OrderBy.cpp (hunk 25 above)
Comment (gggrace14) [reply velox/exec/OrderBy.cpp:94]: Also tried the alternative way to pass RowContainer::estimateRowSize() that I added. The impl of RowContainer::estimateRowSize() is based on allocatedBytes.
https://github.com/facebookincubator/velox/blob/02844168ae12d1a3c6372f2605f1e2e767d8b933/velox/exec/RowContainer.cpp#L564-L575

Here we're at the initializer of OrderBy, and the allocatedBytes is 0 and so estimateRowSize() returns 0. This makes outputBatchRows(RowContainer::estimateRowSize()) returns 1024, kPreferredOutputBatchRows.

For the impl of  RowContainer::estimateRowSize(), I'm wondering if I should return fixedRowSize_, when numRows_ is 0.

Comment (gggrace14) [reply velox/exec/OrderBy.cpp:94]: Per offline discussion, we will change the logic to `outputBatchSize_ = outputBatchRows();`, as the original logic doesn't take into account variable-length rows and perhaps also needs to be improved. Will file an Issue and revisit it with @xiaoxmeng .

File: velox/exec/RowContainer.cpp (hunk 26 above)
Comment (gggrace14) [reply velox/exec/RowContainer.cpp:572]: Revised to check GT

File: velox/exec/tests/RowContainerTest.cpp (hunk 27 above)
Comment (gggrace14) [reply velox/exec/tests/RowContainerTest.cpp:742]: Revised

File: velox/exec/Operator.cpp [hunk 29]
```diff
@@ -294,6 +294,31 @@ OperatorStats Operator::stats(bool clear) {
...
+uint32_t Operator::outputBatchRows(
+    std::optional<uint64_t> averageRowSize) const {
+  const auto& queryConfig = operatorCtx_->task()->queryCtx()->queryConfig();
+
+  if (!averageRowSize.has_value()) {
+    return queryConfig.preferredOutputBatchRows();
+  }
+
+  uint64_t rowSize = averageRowSize.value();
+  VELOX_CHECK_GE(
```
[the remark this answers is not in the document]
Comment (gggrace14) [reply velox/exec/Operator.cpp:306]: I think we shall anticipate the caller to pass averageRowSize 0. I found out from TableScan that averageRowSize passed by caller can be 0.
https://github.com/facebookincubator/velox/blob/0ceaf98faf00717ee66ca75ca273b03f4a6a1034/velox/exec/TableScan.cpp#L259
https://github.com/facebookincubator/velox/blob/9e48ba5dedc53371a506f3ba259dda3728cc9916/velox/dwio/dwrf/reader/DwrfReader.cpp#L453

The condition below this line makes sure that when averageRowSize == 0, this func will return kMaxOutputBatchRows, 10'000. This logic is consistent w/ TableScan. However, I'm wondering if I should change it to kPreferredOutputBatchRows 1024, to be safe. But I think it's okay to take 0 averageRowSize and not throw.

File: velox/exec/OrderBy.cpp [hunk 30]
```diff
@@ -91,12 +91,7 @@ OrderBy::OrderBy(
   }
 #endif
 
-  outputBatchSize_ = std::max<uint32_t>(
```
[the remark this answers is not in the document]
Comment (gggrace14) [reply velox/exec/OrderBy.cpp:94]: Removed it.

I just now found it's also being used in Window.cpp, and I changed it to use outputBatchRows() instead.

However, there seemed to be a bug in Window.cpp originally. If you have time for it, would you possibly help take another look at Window.cpp change, to make sure I'm not breaking it?

File: velox/exec/tests/RowContainerTest.cpp (hunk 28 above)
Comment (gggrace14) [reply velox/exec/tests/RowContainerTest.cpp:754]: I got this idea from TEST_F(rowSizeWithNormalizedKey) and TEST_F(rowSize), and guess it makes sense. As you might know, listRows() returns the first `resultRows` rows that hit the size `rowContainer->estimateRowSize().value() * numRows`. By verifying resultRows == numRows, we see that estimateRowSize() is an accurate estimate.

If there's a better way to verify, would you mind giving an example?

File: velox/exec/Window.cpp [hunk 31]
```diff
@@ -52,8 +52,6 @@ Window::Window(
           operatorId,
           windowNode->id(),
           "Window"),
-      outputBatchSizeInBytes_(
-          driverCtx->queryConfig().preferredOutputBatchSize()),
```
Comment (gggrace14) [inline velox/exec/Window.cpp:56]: @mbasmanova preferredOutputBatchSize() used to mean # of rows, which is 1024. Seems like a bug here. So after my change, behavior will change. But I guess Window op is added recently and not being used a lot.

**Comment by gggrace14** [comment]
Masha, I make the change to have TableScan use it in the PR https://github.com/facebookincubator/velox/pull/4463/commits/b052bacfcbff80ecd7bd927c852edc90a4108ab0, where your comments are kept. TableScan PR is rebased.

File: velox/exec/Operator.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/exec/Operator.cpp:306]: @gggrace14 Ge, thank you for explaining. Maybe document that averageRowSize must be zero or positive and explain the behavior when it is zero.

File: velox/exec/Window.cpp (hunk 31 above)
Comment (mbasmanova) [reply velox/exec/Window.cpp:56]: @gggrace14 Ge, nice catch. 

CC: @aditi-pandit

File: velox/exec/tests/RowContainerTest.cpp (hunk 28 above)
Comment (mbasmanova) [reply velox/exec/tests/RowContainerTest.cpp:758]: @gggrace14 Ge, thank you for explaining. I didn't think about this.

File: velox/exec/RowContainer.cpp [hunk 32]
```diff
@@ -560,6 +561,19 @@ void RowContainer::extractProbedFlags(
   }
 }
 
+std::optional<int64_t> RowContainer::estimateRowSize() const {
+  if (numRows_ == 0) {
+    return std::nullopt;
+  }
+  int64_t freeBytes = rows_.availableInRun() + fixedRowSize_ * numFreeRows_;
```
Comment (xiaoxmeng) [inline velox/exec/RowContainer.cpp:568]:
```
nit:
const int64_t freeBytes
const int64_t usedSize
const int64_t rowSize
```
